flaky: DatasetWidget.drillTitleLabel's identity-keyed case pins a cumulative drawer-render count that overshoots under load

Observed on an unrelated PR (#4705 — plugin-charts doc comment + new plugin-charts test file; it touches no dashboard code). `Test (shard 2/4)` went red on:

```
FAIL  dom  packages/plugin-dashboard/src/__tests__/DatasetWidget.drillTitleLabel.test.tsx
  > DatasetWidget — the drill title reads the series LABEL (objectui#4682) > titles an identity-keyed group with what its segment said, not with its key
AssertionError: expected 2 to be 1 // Object.is equality
 ❯ packages/plugin-dashboard/src/__tests__/DatasetWidget.drillTitleLabel.test.tsx:183:52
```

Job: https://github.com/objectstack-ai/objectui/actions/runs/31885228851/job/95013452561 (shard totals that attempt: `1 failed | 4911 passed`, duration 440.92s — a saturated run). The same run's shards 1/3/4 and every other gate were green.

## Why it is not the PR under test

The failing file is not in #4705's diff, and no cross-file mechanism exists: the `dom` vitest project runs `isolate: true` (per `vitest.config.mts`'s own comment — dom/dom-heavy deliberately keep isolation because they leak otherwise), so the PR's new plugin-charts test file, including its `vi.mock('recharts')`, cannot reach a plugin-dashboard file's module graph. What the PR *did* change is the shard's file composition by one file — enough to shift load and surface a latent race.

## The defect, as the failure text reads

Line 183 is:

```tsx
await waitFor(() => expect(drawerProps.length).toBe(1));
```

`drawerProps` is a recording array of drawer render props, and the pin asserts the **cumulative** count is exactly 1. That assertion can *overshoot*: if the drawer renders twice (an extra state update landing late under load — the recording only ever grows), the observed value is 2 and `waitFor` can never recover, because no amount of waiting shrinks the array. This is the inverse profile of the #4487 family (there the recording was still empty when asserted; here it grew past the pin) — but the same class: a pin bound to incidental scheduling rather than to the property the case is about (per its own title, *which title the drawer shows*, not *how many times React rendered it*).

## Suggested direction (not implemented here)

Assert on the **content of the last recorded props** (the drawer title string the case is actually about), or reset the recording immediately before the drill interaction so exactly-once has a clean window — and if an exactly-once invariant is genuinely intended, pin it against a stable event (e.g. drawer *open* calls), not cumulative renders. Per AGENTS.md 测试纪律, not fixable by raising timeouts.

The test landed with the #4682/#4694 drill-title work earlier today; the sibling cases in the same file share the recording-array idiom and should be swept for the same overshoot exposure by whoever takes this.

## Triage

`pm:queue`: S, same flaky-test class as #4487 (whose fix merged as #4704). It has already red-lighted one unrelated PR on its first day in tree; failed jobs on #4705's head were re-queued as the immediate mitigation.

Provenance: PM review of #4705's shard-2 red, session `session_01RnQd8iMMUwXQEV1crFmQiQ`. Filed unassigned.

--- a/.changeset/dataset-widget-drill-title-render-count.md
+++ b/.changeset/dataset-widget-drill-title-render-count.md
@@ -1,0 +1,9 @@
+---
+---
+
+Test-only: fixed a flaky overshoot in `plugin-dashboard`'s DatasetWidget drill-title
+tests, where `waitFor` pinned a cumulative drawer-render count that a late, unrelated
+dimension-metadata re-render could push past 1 with no way to recover (objectui#4706).
+The assertions now wait for the drawer to have opened at least once and keep their
+content check (the drawer title/filter) as the substantive assertion. No published
+behaviour changes.

--- a/packages/plugin-dashboard/src/__tests__/DatasetWidget.drillTitleLabel.test.tsx
+++ b/packages/plugin-dashboard/src/__tests__/DatasetWidget.drillTitleLabel.test.tsx
@@ -180,7 +180,17 @@ describe('DatasetWidget — the drill title reads the series LABEL (objectui#468
     const ev = clickSegment(1, 0);
     expect(ev.series).toBe('[null]');
 
-    await waitFor(() => expect(drawerProps.length).toBe(1));
+    // Bound to the drawer having opened AT LEAST once (objectui#4706), not to
+    // an exact cumulative render count: `drawerProps` only ever grows, and a
+    // late, unrelated `setMeta` from the dimension-metadata fetch (issued by
+    // `useDatasetDimensionMeta`, resolving independently of the click) can
+    // land a SECOND identical render after this one click — an overshoot
+    // `waitFor` can never recover from once it happens, since the array never
+    // shrinks. The case is about WHICH title the drawer shows, not how many
+    // times React chose to render it, so `lastDrawer()` — the content check
+    // below — is what carries the assertion; the count only needs to confirm
+    // the drawer opened at all.
+    await waitFor(() => expect(drawerProps.length).toBeGreaterThan(0));
     // The KEY did the lookup — the right records, which is what objectui#4673
     // already guaranteed and what must not move…
     expect(lastDrawer().filter).toEqual({ status_id: 'st-backlog', priority_id: null });
@@ -208,7 +218,9 @@ describe('DatasetWidget — the drill title reads the series LABEL (objectui#468
     await waitFor(() => expect(chartSeries().length).toBe(2));
 
     clickSegment(0, 0);
-    await waitFor(() => expect(drawerProps.length).toBe(1));
+    // At-least-once, not exact-once (objectui#4706) — see the first case in
+    // this file for why a cumulative count overshoots.
+    await waitFor(() => expect(drawerProps.length).toBeGreaterThan(0));
     expect(lastDrawer().title).toBe(`Backlog / ${NULL_CATEGORY_LABEL}`);
     expect(lastDrawer().filter).toEqual({ status_id: 'st-backlog', priority_id: 'pri-literal' });
   });
@@ -236,7 +248,9 @@ describe('DatasetWidget — the drill title reads the series LABEL (objectui#468
     ]);
 
     clickSegment(1, 0);
-    await waitFor(() => expect(drawerProps.length).toBe(1));
+    // At-least-once, not exact-once (objectui#4706) — see the first case in
+    // this file for why a cumulative count overshoots.
+    await waitFor(() => expect(drawerProps.length).toBeGreaterThan(0));
     expect(lastDrawer().title).toBe('Backlog / Low');
     expect(lastDrawer().filter).toEqual({ status_id: 'st-backlog', priority_id: 'pri-low' });
   });
@@ -261,7 +275,9 @@ describe('DatasetWidget — the drill title reads the series LABEL (objectui#468
       series: 'High',
     });
 
-    await waitFor(() => expect(drawerProps.length).toBe(1));
+    // At-least-once, not exact-once (objectui#4706) — see the first case in
+    // this file for why a cumulative count overshoots.
+    await waitFor(() => expect(drawerProps.length).toBeGreaterThan(0));
     expect(lastDrawer().title).toBe('Backlog / High');
   });
 
@@ -303,7 +319,9 @@ describe('DatasetWidget — the drill title reads the series LABEL (objectui#468
       seriesLabel: 'Hours',
     });
 
-    await waitFor(() => expect(drawerProps.length).toBe(1));
+    // At-least-once, not exact-once (objectui#4706) — see the first case in
+    // this file for why a cumulative count overshoots.
+    await waitFor(() => expect(drawerProps.length).toBeGreaterThan(0));
     expect(lastDrawer().title).toBe('Backlog');
   });
 });

